### PR TITLE
Hulto eldritch feature pivot ncat

### PR DESCRIPTION
**Pull request by hulto**
#### What this PR does / why we need it:

This PR adds support for eldritch to send UDP / TCP data to an arbitrary host / port. If the host/port responds with any data that data will be returned to the user.

### CODE REVIEW

**Comment by hulto** [comment]
This PR also includes stubs for the pivot library.

File: implants/eldritch/src/pivot/ncat_impl.rs [hunk 1]
```diff
@@ -0,0 +1,172 @@
+
+use std::net::Ipv4Addr;
+
+use anyhow::Result;
+use tokio::io::{AsyncWriteExt, AsyncReadExt};
+use tokio::net::{TcpStream, UdpSocket};
+
+// Since we cannot go from async (test) -> sync (ncat) `block_on` -> async (handle_ncat) without getting an error "cannot create runtime in current runtime since current thread is calling async code."
+async fn handle_ncat(address: String, port: i32, data: String, protocol: String) -> Result<String> {
```
Comment (hulto) [inline implants/eldritch/src/pivot/ncat_impl.rs:9]: The working portion and the primary function have been separated to make testing async code easier.

File: implants/eldritch/src/pivot/ncat_impl.rs [hunk 2]
```diff
@@ -0,0 +1,177 @@
...
+    // If the response is longer than 4096 bytes it will be  truncated.
+    let mut response_buffer = [0; 4096];
```
Comment (hulto) [inline implants/eldritch/src/pivot/ncat_impl.rs:11]: Not sure if there's a better way to receive more bytes.
If not should we add a truncated indicator? Eg. "I got. lots of bytes of data ...trunc...

Comment (Milkshak3s) [reply implants/eldritch/src/pivot/ncat_impl.rs:11]: Can we use `read_exact` either to read enough header to determine the packet size, or alternatively read_exact in a loop?

https://docs.rs/tokio/0.1.22/tokio/io/fn.read_exact.html

Comment (hulto) [reply implants/eldritch/src/pivot/ncat_impl.rs:11]: Good suggestion.
Do you know of a good way to read an undefined number of bytes? I think this still uses a fixed length buffer.

File: implants/eldritch/src/pivot/ncat_impl.rs [hunk 3]
```diff
@@ -1,5 +1,209 @@
...
+        // We  need to take a buffer of bytes, turn it into a String but that string has null bytes.
+        // To remove the null bytes we're using trim_matches.
+        result_string = String::from(String::from_utf8((&response_buffer).to_vec()).unwrap().trim_matches(char::from(0)));
+        Ok(result_string)
+
+    } else if protocol == "udp" {
+        // Connect to remote host
+    
+        // Setting the bind address to unspecified should leave it up to the OS to decide.
+        // https://stackoverflow.com/a/67084977
```
Comment (Milkshak3s) [inline implants/eldritch/src/pivot/ncat_impl.rs:35]: 🔥

File: implants/eldritch/src/pivot/ncat_impl.rs [hunk 4]
```diff
@@ -1,5 +1,209 @@
...
+        // To remove the null bytes we're using trim_matches.
+        result_string = String::from(String::from_utf8((&response_buffer).to_vec()).unwrap().trim_matches(char::from(0)));
+        Ok(result_string)
+
+    } else {
+        return Err(anyhow::anyhow!("Protocol not supported please use: udp or tcp."));
+    }
+}
+
+// We do not want to make this async since it would require we make all of the starlark bindings async.
```
Comment (Milkshak3s) [inline implants/eldritch/src/pivot/ncat_impl.rs:55]: Good to know :blob_salute:

**Review by Milkshak3s — APPROVED** [review]
LGTM!

**Pull request merged**